### DISCUSSION TITLE
postCreateCommand weirdness

### DISCUSSION THREAD

**Discussion by makestuff** [community/community#27133] [Codespaces]
<p>Hi!</p>
<p>I’ve been experimenting with GitHub Codespaces, and I encountered a problem with <code>postCreateCommand</code>. I’d like it to run the <code>build.sh</code> script from <a href="https://github.com/makestuff/vscode-foo" rel="noopener nofollow ugc">my repo</a>, with a couple of arguments, so in my <code>.devcontainer/devcontainer.json</code> file, I did <a href="https://github.com/makestuff/vscode-foo/blob/main/.devcontainer/devcontainer.json#L57" rel="noopener nofollow ugc">this</a>:</p>
<pre><code>"postCreateCommand": "./build.sh Debug -nobuild"
</code></pre>
<p>The outcome is:</p>
<pre><code>Starting Codespace...
2021-01-18T19:21:28: $ docker exec -t -u vscode -w /home/vscode/workspace/vscode-foo -e GITHUB_TOKEN="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" -e GITHUB_CODESPACE_TOKEN="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" -e GIT_COMMITTER_NAME="GitHub" -e GIT_COMMITTER_EMAIL="noreply@github.com" 4ccd05e577f6 sh -c \"'./build.sh Debug -nobuild'\"
2021-01-18T19:21:28: Debug: 1: Debug: Syntax error: Unterminated quoted string
2021-01-18T19:21:28: docker process exited with exit code 2
</code></pre>
<p>Note the command executed: <code>sh -c \"'./build.sh Debug -nobuild'\"</code>. Note that the command is wrapped by <code>\"</code>. That’s not correct.</p>
<p>I can do a local build (i.e on my PC) of the image specified by my <code>Dockerfile</code>, then run the image with a <code>git clone</code> of my repo, to replicate the state of the codespace:</p>
<pre><code>host$ cd $HOME
host$ git clone https://github.com/makestuff/vscode-foo.git
        :
host$ cd vscode-foo/.devcontainer
host$ docker build --tag makestuff:0.1 .
      :
host$ docker run --name 4ccd05e577f6 --user vscode -it makestuff:0.1 bash
vscode ➜ / $ cd /home/vscode &amp;&amp; mkdir -p workspace &amp;&amp; cd workspace &amp;&amp; git clone --recursive https://github.com/makestuff/vscode-foo.git
  :
vscode ➜ ~/workspace $
</code></pre>
<p>Once done, in order to get a sane response to the <code>docker exec</code> above, I need to do it without the surrounding <code>\"</code> quotes:</p>
<pre><code>docker exec -t -u vscode -w /home/vscode/workspace/vscode-foo -e GITHUB_TOKEN="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" -e GITHUB_CODESPACE_TOKEN="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" -e GIT_COMMITTER_NAME="GitHub" -e GIT_COMMITTER_EMAIL="noreply@github.com" 4ccd05e577f6 sh -c './build.sh Debug -nobuild'
</code></pre>
<p>Note the command executed: <code>sh -c './build.sh Debug -nobuild'</code>. That works fine.</p>
<p>Why is the value I give for <code>postBuildCommand</code> being wrapped in <code>\"</code> in this way?</p>
<p>Thanks!</p>
<p>- Chris</p>

**Comment by freekboutkan-ah** [comment]
<aside class="quote no-group" data-username="makestuff" data-post="1" data-topic="156256">
<div class="title">
<div class="quote-controls"></div>
 makestuff:</div>
<blockquote>
<p><code>8: Debug: 1: Debug: Syntax error: Unterminated quoted string</code></p>
</blockquote>
</aside>
<p>I have the exact same issue. I have defined this is my devcontainer.json file<br>
<code>"postCreateCommand": "pip3 --no-cache-dir install -e .[local]"</code></p>
<p>In the creation logs I see:</p>
<pre><code class="lang-auto">Starting Codespace...
2021-01-18T18:20:58: $ docker exec -t -u vscode -w /home/vscode/workspace/*** -e GITHUB_TOKEN="***" -e GITHUB_CODESPACE_TOKEN="***" -e GIT_COMMITTER_NAME="GitHub" -e GIT_COMMITTER_EMAIL="noreply@github.com" *** sh -c \"'pip3 --no-cache-dir install -e .[local]'\"

2021-01-18T18:20:58: --no-cache-dir: 1: --no-cache-dir: Syntax error: Unterminated quoted string

2021-01-18T18:20:58: docker process exited with exit code 2
</code></pre>

**Comment by arcdigital** [comment]
<p>I’m having the same issue.</p>

**Comment by jarrodldavis** [comment]
<p>I’m also having issues with a <code>postCreateCommand</code> that works fine locally (using the Remote - Containers extension) doesn’t work in GitHub Codespaces. I either get a “command not found” error or a generic “completed with errors” error.</p>

**Comment by jarrodldavis** [comment]
<p>Quick update to this, something broken definitely got deployed recently. Even <code>pwd</code> doesn’t work:</p>
<pre><code class="lang-auto">Starting Codespace...

[Run container]

Configuring codespace...
$ docker exec -t -u vscode a3aee51a3c39 echo -e 

docker process exited with exit code 0
$ docker exec -t -u vscode a3aee51a3c39 echo -e 
[Configure codespace]


docker process exited with exit code 0
$ docker exec -t -u vscode -w /home/vscode/workspace/solomon -e GITHUB_TOKEN="&lt;redacted&gt;" -e GITHUB_CODESPACE_TOKEN="&lt;redacted&gt;" -e GIT_COMMITTER_NAME="GitHub" -e GIT_COMMITTER_EMAIL="noreply@github.com" a3aee51a3c39 sh -c \"'pwd'\"
sh: 1: 'pwd': not found
docker process exited with exit code 127
</code></pre>
<hr>
<p>Also, the typical behavior where an array-valued <code>postCreateCommand</code> is ran directly instead of through a shell isn’t working either. This is how I have my <code>devcontainer.json</code>:</p>
<pre><code class="lang-json">  // Use 'postCreateCommand' to run commands after the container is created.
  "postCreateCommand": ["pwd"],
</code></pre>
<p>That should run <code>pwd</code> (i.e. <code>/bin/pwd</code>) directly instead of running it under using <code>sh</code>.</p>

**Comment by makestuff** [comment]
<p>OK, I give up. How are we supposed to report bugs with GitHub Codespaces? The most promising link I could find was “<a href="https://docs.github.com/en/github/developing-online-with-codespaces/about-codespaces#contacting-us-about-codespaces" rel="noopener nofollow ugc">Contacting us about Codespaces</a>”, which says:</p>
<blockquote>
<p>If you encounter problems using Codespaces, see “<a href="https://docs.github.com/en/github/developing-online-with-codespaces/troubleshooting-your-codespace" rel="noopener nofollow ugc">Troubleshooting your codespace</a>”</p>
</blockquote>

**Comment by makestuff** [comment]
<p>…which in turn says:</p>
<blockquote>
<p>If you still need help, you can contact us. For more information, see “<a href="https://docs.github.com/en/github/developing-online-with-codespaces/about-codespaces#contacting-us-about-codespaces" rel="noopener nofollow ugc">About Codespaces</a>.”</p>
</blockquote>
<p>…which results in an infinite loop.</p>
<p>(Sorry about splitting this into two posts - the forum won’t let me post more than two links)</p>

**Comment by jarrodldavis** [comment]
<p>Looks like that also says to use the Community Forums, so I guess you’re in the right place. Now we just wait until a GitHub employee shows up.</p>

**Comment by jkeech** [comment]
<p>Hey everyone, thanks for reporting this issue. We’re actively investigating <code>postCreateCommand</code> not executing for new codespaces.</p>

**Comment by jkeech** [comment, marked as answer]
<p>Apologies for the issue, this issue should be fixed now for all new Codespaces.</p>

**Comment by makestuff** [comment]
<p>Amazing, thanks for the quick fix, <a class="mention" href="/jkeech">@jkeech</a>. I can confirm this is working now.</p>
